qcodes/tests/dataset/test_measurement_context_manager.py

```diff
@@ -713,6 +714,68 @@ def sub_get_y_vals(results, length, state):
     assert yvals == list(given_yvals)
 
 
+@settings(max_examples=5, deadline=None)
+@given(N=hst.integers(min_value=100, max_value=200))
+def test_example_subscriber_text_progress(capsys, experiment, DAC, DMM, N):
+    """
+    Test the TextProgress example subscriber.
+    """
+
+    def do_test(variant: int = 0):
+        # variant 0: without setting total_measurements
+        # variant 1: set total_measurements as parameter
+        # variant 2: set total_measurements via attribute
+
+        meas = Measurement(exp=experiment)
+        meas.register_parameter(DAC.ch1)
+        meas.register_parameter(DMM.v1, setpoints=(DAC.ch1,))
+
+        if variant == 0:
+            meas.add_subscriber(TextProgress(), state=None)
+        elif variant == 1:
+            meas.add_subscriber(TextProgress(total_measurements=N), state=None)
+        elif variant == 2:
+            tp = TextProgress()
+            meas.add_subscriber(tp, state=None)
+        else:
+            raise ValueError(f'Invalid variant {variant}')
+
+        with meas.run() as datasaver:
+
+            if variant == 2:
+                tp.total_measurements = N
+
+            capsys.readouterr()  # flush stdout/stderr
+
+            for i, (x, y) in enumerate(zip(range(N), range(1, N+1))):
+                datasaver.add_result((DAC.ch1, x), (DMM.v1, y))
+
+                datasaver.flush_data_to_database()
+
+                # we assume that the test takes less than a second
+                ptn = None
+                if variant == 0:
+                    ptn = rf'{i + 1} - elapsed: 0:00:0\d'
+                else:
+                    percent = f'{(i + 1) / N * 100 :.1f} %'
+                    ptn = rf'{i + 1} / {N} \({percent}\) - elapsed: 0:00:0\d'
+                    if i > 0:
+                        ptn += r' - ETA: 0:00:0\d'
+                        if variant == 2:
+                            ptn += r' \(\w+ \w+ \d\d \d\d:\d\d:\d\d\)'
+                        elif variant == 3:
+                            ptn += r' \(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d\)'
```

Comment:
this is an impressive amount of effort, to assert stdout with regexps. However, i think it's more useful to inject something into `TextProgress` which would just collect all the "messages" that it wants to print to the screen in some structured format (for example, a dictionary), and then the test would assert the contents of that dictionary. Such an injection in the dirty way can be done using magic mock (i'd recommend to rewrite the `__call__` method to create such a structured "message" and then pass it to some `_print_progress_message` method, which can be magic-mocked to also store the given "messages" somewhere where they can be asserted by the test method). I also do think that despite the use of a mock or similar things, the test readability will increase (as opposed to trying to make sense of the regexps).